### PR TITLE
Allow an Override Flag in UsingTasks

### PR DESCRIPTION
**Pull request by BenVillalobos**
Fixes https://github.com/dotnet/msbuild/issues/5541

### Context
When we define multiple usingtasks like so:
```xml
  <UsingTask TaskName="VSCTCompiler" AssemblyFile="$(VsSDKAssemblyFile)" />
  <UsingTask TaskName="VSCTCompiler" AssemblyFile="$(VsSDKAssemblyFile)" Architecture="x86" />
```

and call the task without specifying `Architecture`, the architecture-specific `UsingTask` should be returned when MSBuild searches for the task. Currently, MSBuild prefers the most "exact match".

This change allows the following:
```xml
  <UsingTask TaskName="VSCTCompiler" AssemblyFile="$(VsSDKAssemblyFile)" />
  <UsingTask TaskName="VSCTCompiler" AssemblyFile="$(VsSDKAssemblyFile)" Override="true" Architecture="x86" />
```
Users can specify `Override="true"` in a `UsingTask` element. This will modify the TaskRegistry to _always return that task registration_ when attempting to call that task.

### Changes Made
The TaskRegistry will return the first usingtask marked with `Override="true"` when searching for a task.

### Testing
A test showing that:

- Multiple usingtasks overriding the same task will log MSB4275
- If a usingtask overrides, it is preferred over other usingtasks.

### Notes
Don't review commit by commit.

# Expected behaviors
### Prioritization order of usingtasks:
1. Project-level usingtask override
2. *.tasks/*.overridetasks file
3. UsingTask defined elsewhere

### CODE REVIEW

**Comment by BenVillalobos** [comment]
Thinking out loud, is #5541 more about when a task has `Architecture` defined? Maybe we can add a check that prioritizes the usingtask that has `Architecture` (and any other "high priority" attributes) defined when trying to find the task from the registry?

**Comment by rainersigwald** [comment]
> Thinking out loud, is #5541 more about when a task has `Architecture` defined?

I think so; the question is "given an old poorly-annotated x86 task + target buried in a NuGet package or old SDK or something, is there something a user project can do to fix the build?" With other things (like targets or properties or items) you can override behavior if you can get the ordering right. But here you can't.

> Maybe we can add a check that prioritizes the usingtask that has `Architecture` (and any other "high priority" attributes) defined when trying to find the task from the registry?

I think so? Architecture is the only one we care about right now, maybe someday we'll want core/framework.

**Comment by BenVillalobos** [comment]
Current thought for this PR is to have some Dictionary containing <Name of task that has an Architecture-specific usingtask, The registration record for that task>. And I'll short-circuit the task lookup method when the task we're looking for happens to have an architecture-specific version.

**Comment by BenVillalobos** [comment]
NTS: I didn't consider there potentially being multiple usingtasks that define Architecture but I haven't gotten to the selection process for preferred usingtasks yet. GenerateResource so far has multiple usingtasks.

**Comment by BenVillalobos** [comment]
> NTS: I didn't consider there potentially being multiple usingtasks that define Architecture but I haven't gotten to the selection process for preferred usingtasks yet. GenerateResource so far has multiple usingtasks.

From standup today, if multiple UsingTasks define `Architecture`, first one wins.

**Comment by BenVillalobos** [comment]
NTS: But what about different tasks with the same name? Might need to check the dll its coming from as well.

**Comment by BenVillalobos** [comment]
Because this will be modifying behavior of `UsingTasks`, we should update https://docs.microsoft.com/en-us/visualstudio/msbuild/usingtask-element-msbuild?view=vs-2019 when this merges.

**Comment by BenVillalobos** [comment]
NTS: Need to modify tests in such a way that account for _any_ usingtask with Architecture defined is prioritized. The problem is _many_ tests check for Architecture/Runtime when doing their tests. Might be able to get away without the bool "hack".

**Comment by BenVillalobos** [comment]
Notes after chatting with @rainersigwald
----
Keep in mind that we'll want to generalize this for `Runtime` as a parameter as well.

The arch/runtime should match the current arch/runtime.

Final thoughts we came to: A usingtask should specify some `Override="true"` to opt into this. If that's the case, it should override ANY call to a task with that name. **Even if there are two tasks with the same name from two separate assemblies.** At that point a user can call `Namespace.Class.Foo` to call the specific task they want.

We were looking at a project with the following usingtasks:

```xml
  
<UsingTask TaskName="Foo" AssemblyFile="$(Outdir)task.dll" Architecture="x86" />
<UsingTask TaskName="Foo" AssemblyFile="$(Outdir)task.dll"/>
<UsingTask TaskName="Foo" AssemblyFile="$(Outdir)task234.dll"/>
<!-- Problem: If Foo is called, the second declaration here will be 
       found on "Exact" match before fuzzy search happens. -->

  <Target Name="Foo" AfterTargets="Build">
    <Foo/>
  </Target>

```

The first UsingTask will be stored in a dictionary of <(TaskName, TaskIdentity), List<Assemblies>>.

High level ordering of relevant functions when finding a task with a given name:

- FindTaskInRegistry <- First attempts exact match then fuzzy
	- GetRegisteredTask
		- GetTaskRegistrationRecord
			- // If fuzzy match
				- GetRelevantRegistrations <- Returns Dictionary<TaskIdentity, List<RegisteredTaskRecord>>
					- for each registration
						- GetMatchingRegistration <- Finds the **first partial match**

We're thinking the bulk of the change should be made in `GetRelevantRegistrations` or `RegisterTask`.

Options if we modify GetRelevantRegistrations:
- Sort usingtasks based on this Override flag, always return that one.
- Store tasks marked as Override in a separate collection, short circuit this method if the task being searched for has an `Override` usingtask defined.

Options if we modify RegisterTask:
- If some task `Foo` has `Override` specified, delete or modify existing entries of the `Foo` task to specifically use this usingtask.

File: src/Build/Instance/TaskRegistry.cs [hunk 1]
```diff
@@ -411,10 +412,10 @@ ElementLocation elementLocation
         )
         {
             TaskFactoryWrapper taskFactory = null;
-            bool retrievedFromCache;
-
+            bool retrievedFromCache = false;
+            
             // If there are no usingtask tags in the project don't bother caching or looking for tasks locally
-            RegisteredTaskRecord record = GetTaskRegistrationRecord(taskName, taskProjectFile, taskIdentityParameters, exactMatchRequired, targetLoggingContext, elementLocation, out retrievedFromCache);
+            RegisteredTaskRecord record = overriddenTasks.ContainsKey(taskName) ? overriddenTasks[taskName] : GetTaskRegistrationRecord(taskName, taskProjectFile, taskIdentityParameters, exactMatchRequired, targetLoggingContext, elementLocation, out retrievedFromCache);
```
Comment (Forgind) [inline src/Build/Instance/TaskRegistry.cs:418]: Right spot?

Comment (BenVillalobos) [inline src/Build/Instance/TaskRegistry.cs:418]: Note for reviewers: Is this the right spot to be returning early?

File: src/Build/Instance/TaskRegistry.cs [hunk 2]
```diff
@@ -471,6 +472,12 @@ internal RegisteredTaskRecord GetTaskRegistrationRecord
             retrievedFromCache = false;
             RegisteredTaskIdentity taskIdentity = new RegisteredTaskIdentity(taskName, taskIdentityParameters);
 
+            // was this task overridden?
+            if (overriddenTasks.TryGetValue(taskName, out RegisteredTaskRecord rec))
```
Comment (BenVillalobos) [inline src/Build/Instance/TaskRegistry.cs:476]: Note: I moved this check down 1 layer from `GetRegisteredTask` to `GetTaskRegistrationRecord` since the former calls the latter no matter what.

Unit tests also call directly into the latter.

**Review by baronfel — COMMENTED** [review]
The proposed design makes sense to me, a few questions I had are:

* would a user ever want to know about this decision being made?  if so, should there be a reason why the decision was made? e.g a `OverrideReason` attribute where inline doc could be added, so we could add a `MS1000000: Task 'TaskName' from assembly 'AssemblyName' was chosen instead of 'other task options' because of 'OverrideReason'` informational-level warning.
* the new field should be added to the XSD with brief docs
*

**Comment by Forgind** [comment]
Informational level message you mean? I think "Override" is clear enough that we shouldn't give a warning.

**Comment by baronfel** [comment]
Yeah, an informational level message is the correct term.

File: src/Build.UnitTests/BackEnd/TaskRegistry_Tests.cs [hunk 3]
```diff
@@ -622,6 +622,99 @@ public void RetrieveFromCacheMatchingExactParameters_AdditionalParameters()
                 );
         }
 
+        [Fact]
+        public void OverriddenTask_AlwaysWins()
+        {
```
Comment (Forgind) [inline src/Build.UnitTests/BackEnd/TaskRegistry_Tests.cs:627]: Can you combine these two Facts into a Theory then add a case for element having Architecture = "x64" and Override set to "false" then secondElement having Architecture "x86" and Override "true"?

File: src/Build/Instance/TaskRegistry.cs [hunk 4]
```diff
@@ -411,8 +412,8 @@ ElementLocation elementLocation
         )
         {
             TaskFactoryWrapper taskFactory = null;
-            bool retrievedFromCache;
-
+            bool retrievedFromCache = false;
```
Comment (Forgind) [inline src/Build/Instance/TaskRegistry.cs:415]: Why this change? You can inline it instead if you want with ...out bool retrievedFromCache);

File: src/Build/Instance/TaskRegistry.cs [hunk 5]
```diff
@@ -660,7 +671,14 @@ private void RegisterTask(string taskName, AssemblyLoadInfo assemblyLoadInfo, st
                 _taskRegistrations[taskIdentity] = registeredTaskEntries;
             }
 
-            registeredTaskEntries.Add(new RegisteredTaskRecord(taskName, assemblyLoadInfo, taskFactory, taskFactoryParameters, inlineTaskRecord));
+            RegisteredTaskRecord newRecord = new RegisteredTaskRecord(taskName, assemblyLoadInfo, taskFactory, taskFactoryParameters, inlineTaskRecord);
+
+            if (overrideTask && !overriddenTasks.ContainsKey(taskName))
```
Comment (Forgind) [inline src/Build/Instance/TaskRegistry.cs:676]: It seems like you should be able to override an override.
```suggestion
            if (overrideTask)
```
Maybe a warning?

**Review by Forgind — APPROVED** [review]
Looks roughly good, though I'm a little worried that you changed TaskRegistry_Tests so much.

File: src/Build.UnitTests/BackEnd/TaskRegistry_Tests.cs [hunk 6]
```diff
@@ -651,7 +744,7 @@ public void RetrieveFromCacheFuzzyMatchingParameters()
                     shouldBeRetrieved: true,
                     shouldBeRetrievedFromCache: false,
                     expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.clr4,
-                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.x86
```
[the remark this answers is not in the document]
Comment (BenVillalobos) [reply src/Build.UnitTests/BackEnd/TaskRegistry_Tests.cs:654]: Ah these are holdovers from the previous solution. Will need to revert the commit that changed all these tests.

**Comment by BenVillalobos** [comment]
> would a user ever want to know about this decision being made? if so, should there be a reason why the decision was made? e.g a OverrideReason attribute where inline doc could be added, so we could add a MS1000000: Task 'TaskName' from assembly 'AssemblyName' was chosen instead of 'other task options' because of 'OverrideReason' informational-level warning.

A message sounds reasonable when registering an overridden task. Related to another comment in this thread, should we log a warning when multiple usingtasks are marked as `Override=true` (with the same task name)?

File: src/Build/Instance/TaskRegistry.cs (hunk 5 above)
Comment (baronfel) [reply src/Build/Instance/TaskRegistry.cs:675]: When would this happen? Some complex `Condition` evaluation that a user thought was exhaustive but didn't quite catch one particular corner case?  It's my understanding that most `<UsingTask>`s are done in a block, with conditions and such to make sure that the best one is chosen. That would mean multiple `Overrides` applying would be unexpected.

**Comment by baronfel** [comment]
> 
> 
> > would a user ever want to know about this decision being made? if so, should there be a reason why the decision was made? e.g a OverrideReason attribute where inline doc could be added, so we could add a MS1000000: Task 'TaskName' from assembly 'AssemblyName' was chosen instead of 'other task options' because of 'OverrideReason' informational-level warning.
> 
> A message sounds reasonable when registering an overridden task. Related to another comment in this thread, should we log a warning when multiple usingtasks are marked as `Override=true` (with the same task name)?

I think yes, because at that point we had to make a judgement call (first seen, last seen, random based on entropy seed generated from a bee's wings) as to which Task's configuration was actually used. I think anytime we have to make a judgement call like that, the user should be told about it.

**Comment by BenVillalobos** [comment]
I think the current test failure has to do with the build context when calling this task directly? I haven't seen an issue like this before. Will try to check for the logged warning case by building a full project that has multiple usingtasks in it, instead of calling the method directly.

File: src/Build/Instance/TaskRegistry.cs [hunk 7]
```diff
@@ -660,7 +681,22 @@ private void RegisterTask(string taskName, AssemblyLoadInfo assemblyLoadInfo, st
...
             }
 
-            registeredTaskEntries.Add(new RegisteredTaskRecord(taskName, assemblyLoadInfo, taskFactory, taskFactoryParameters, inlineTaskRecord));
+            RegisteredTaskRecord newRecord = new RegisteredTaskRecord(taskName, assemblyLoadInfo, taskFactory, taskFactoryParameters, inlineTaskRecord);
+
+            if (overrideTask)
+            {
+                if (overriddenTasks.ContainsKey(taskName))
+                {
+                    loggingService.LogWarning(context, null, new BuildEventFileInfo(projectUsingTaskInXml.OverrideLocation), "DuplicateOverrideUsingTaskElement", taskName);
```
Comment (Forgind) [inline src/Build/Instance/TaskRegistry.cs:690]: I'm split but leaning slightly towards that you should always override it, so essentially moving overridenTasks[taskName] = newRecord; out of this if/else part. Warning/message difference should still exist though.

**Review by Forgind — APPROVED** [review]
Tests look much better, thanks

File: src/Build/Resources/Strings.resx [hunk 8]
```diff
@@ -331,6 +331,18 @@
   </data>
   <data name="TryingExtensionsPath" xml:space="preserve">
     <value>Trying to import {0} using extensions path {1}</value>
+  </data>
+    <data name="OverrideUsingTaskElementCreated" xml:space="preserve">
```
Comment (Forgind) [inline src/Build/Resources/Strings.resx:335]: Why do all of these need error codes? I would've thought just the DuplicateOverrideUsingTaskElement should.

File: src/Build/Instance/TaskRegistry.cs [hunk 9]
```diff
@@ -635,11 +642,25 @@ private static bool IsTaskFactoryClass(Type type, object unused)
...
         /// </summary>
-        private void RegisterTask(string taskName, AssemblyLoadInfo assemblyLoadInfo, string taskFactory, Dictionary<string, string> taskFactoryParameters, RegisteredTaskRecord.ParameterGroupAndTaskElementRecord inlineTaskRecord)
+        private void RegisterTask
+        (
+            string taskName,
+            AssemblyLoadInfo assemblyLoadInfo,
+            string taskFactory,
+            Dictionary<string, string> taskFactoryParameters,
+            RegisteredTaskRecord.ParameterGroupAndTaskElementRecord inlineTaskRecord,
+            ILoggingService loggingService, BuildEventContext context,
```
Comment (Forgind) [inline src/Build/Instance/TaskRegistry.cs:660]: Split onto two lines

File: ref/Microsoft.Build/net/Microsoft.Build.cs [hunk 10]
```diff
@@ -412,6 +412,8 @@ public partial class ProjectUsingTaskElement : Microsoft.Build.Construction.Proj
         public Microsoft.Build.Construction.ElementLocation AssemblyFileLocation { get { throw null; } }
         public string AssemblyName { get { throw null; } set { } }
         public Microsoft.Build.Construction.ElementLocation AssemblyNameLocation { get { throw null; } }
+        public string Override { get { throw null; } set { } }
+        public Microsoft.Build.Construction.ElementLocation OverrideLocation { get { throw null; } }
```
Comment (rainersigwald) [inline ref/Microsoft.Build/net/Microsoft.Build.cs:416]: Can you rebase on main to pick up the PublicApiAnalyzer? that'll force you to rebuild this stuff.

File: src/Build/Resources/Strings.resx [hunk 11]
```diff
@@ -331,6 +331,18 @@
...
+    <value>MSB4275: Created an override using task: {0}</value>
+    <comment>{StrBegin="MSB4275: "}</comment>
+  </data>
+  <data name="DuplicateOverrideUsingTaskElement" xml:space="preserve">
+    <value>MSB4276: Attempted to create multiple overrides of the same task: {0}</value>
```
Comment (rainersigwald) [inline src/Build/Resources/Strings.resx:340]: Is there a test that this gets thrown? I didn't see one.

File: src/Build/Instance/TaskRegistry.cs [hunk 12]
```diff
@@ -386,7 +387,7 @@ IFileSystem fileSystem
                 taskFactoryParameters.Add(XMakeAttributes.architecture, architecture == String.Empty ? XMakeAttributes.MSBuildArchitectureValues.any : architecture);
             }
 
-            taskRegistry.RegisterTask(taskName, AssemblyLoadInfo.Create(assemblyName, assemblyFile), taskFactory, taskFactoryParameters, parameterGroupAndTaskElementRecord);
+            taskRegistry.RegisterTask(taskName, AssemblyLoadInfo.Create(assemblyName, assemblyFile), taskFactory, taskFactoryParameters, parameterGroupAndTaskElementRecord, loggingService, buildEventContext, projectUsingTaskXml, overrideUsingTask.Equals("true", StringComparison.OrdinalIgnoreCase));
```
Comment (rainersigwald) [inline src/Build/Instance/TaskRegistry.cs:390]:
```suggestion
            taskRegistry.RegisterTask(taskName, AssemblyLoadInfo.Create(assemblyName, assemblyFile), taskFactory, taskFactoryParameters, parameterGroupAndTaskElementRecord, loggingService, buildEventContext, projectUsingTaskXml, ConversionUtilities.ValidBooleanTrue(overrideUsingTask));
```

(this handles more cases including some terrifying ones)

File: src/Build/Resources/Strings.resx [hunk 13]
```diff
@@ -331,6 +331,18 @@
...
+    <comment>{StrBegin="MSB4276: "}</comment>
+  </data>
+    <data name="OverriddenUsingTaskElementChosen" xml:space="preserve">
+    <value>MSB4277: Overridden task chosen when calling task: {0}</value>
```
Comment (rainersigwald) [inline src/Build/Resources/Strings.resx:344]: Should this have a pointer to the XML location of the override? If it's wrong (which is the only time I can think of to care about this message) that's probably interesting.

File: src/Build/Resources/Strings.resx [hunk 14]
```diff
@@ -331,6 +331,18 @@
   </data>
   <data name="TryingExtensionsPath" xml:space="preserve">
     <value>Trying to import {0} using extensions path {1}</value>
+  </data>
+    <data name="OverrideUsingTaskElementCreated" xml:space="preserve">
+    <value>MSB4275: Created an override using task: {0}</value>
```
Comment (rainersigwald) [inline src/Build/Resources/Strings.resx:336]: Do we need _both_ this and the "chosen" one? I feel like one or the other would do.

File: src/Build/Instance/TaskRegistry.cs [hunk 15]
```diff
@@ -660,7 +682,22 @@ private void RegisterTask(string taskName, AssemblyLoadInfo assemblyLoadInfo, st
                 _taskRegistrations[taskIdentity] = registeredTaskEntries;
             }
 
-            registeredTaskEntries.Add(new RegisteredTaskRecord(taskName, assemblyLoadInfo, taskFactory, taskFactoryParameters, inlineTaskRecord));
+            RegisteredTaskRecord newRecord = new RegisteredTaskRecord(taskName, assemblyLoadInfo, taskFactory, taskFactoryParameters, inlineTaskRecord);
```
Comment (rainersigwald) [inline src/Build/Instance/TaskRegistry.cs:685]: The cool kids are doing it this way now (I don't actually care which you do):

```suggestion
            RegisteredTaskRecord newRecord = new(taskName, assemblyLoadInfo, taskFactory, taskFactoryParameters, inlineTaskRecord);
```

File: src/Build/Instance/TaskRegistry.cs [hunk 16]
```diff
@@ -635,11 +642,26 @@ private static bool IsTaskFactoryClass(Type type, object unused)
             return relevantTaskRegistrations;
         }
 
+        // Create another set containing architecture-specific task entries.
+        // Then when we look for them, check if the name exists in that.
+        Dictionary<string, RegisteredTaskRecord> overriddenTasks = new Dictionary<string, RegisteredTaskRecord>();
```
Comment (rainersigwald) [inline src/Build/Instance/TaskRegistry.cs:647]: You're keying off of a `string` name--how does that work with the partial-class-name resolution that normal task lookup does? Does this require an exact match to the invocation?

File: src/Build/Instance/TaskRegistry.cs (hunk 7 above)
Comment (rainersigwald) [reply src/Build/Instance/TaskRegistry.cs:691]: That would make it last-one-wins, where normal behavior for UsingTasks is first-one-wins, so I think I prefer the current approach.

File: src/Build/Instance/TaskRegistry.cs [hunk 17]
```diff
@@ -660,7 +682,22 @@ private void RegisterTask(string taskName, AssemblyLoadInfo assemblyLoadInfo, st
...
             }
 
-            registeredTaskEntries.Add(new RegisteredTaskRecord(taskName, assemblyLoadInfo, taskFactory, taskFactoryParameters, inlineTaskRecord));
+            RegisteredTaskRecord newRecord = new RegisteredTaskRecord(taskName, assemblyLoadInfo, taskFactory, taskFactoryParameters, inlineTaskRecord);
+
+            if (overrideTask)
+            {
+                if (overriddenTasks.ContainsKey(taskName))
+                {
+                    loggingService.LogWarning(context, null, new BuildEventFileInfo(projectUsingTaskInXml.OverrideLocation), "DuplicateOverrideUsingTaskElement", taskName);
```
Comment (rainersigwald) [inline src/Build/Instance/TaskRegistry.cs:691]: Why a warning and not an error?

**Review by rainersigwald — COMMENTED** [review]
Can you also prepare a draft of the docs changes that will describe this? I think that'll help clarify the behavior design.

File: src/Build/Instance/TaskRegistry.cs (hunk 7 above)
Comment (Forgind) [reply src/Build/Instance/TaskRegistry.cs:691]: I might argue that we don't have a default yet for overriding usingtasks, and this is more similar to how we do things in other parts of MSBuild, but I'm fine with this.

File: src/Build/Resources/Strings.resx (hunk 11 above)
Comment (BenVillalobos) [reply src/Build/Resources/Strings.resx:336]: Just added it :)

File: src/Build/Instance/TaskRegistry.cs (hunk 12 above)
Comment (Forgind) [reply src/Build/Instance/TaskRegistry.cs:390]: Do we have any data on how often some of the less-used "valid booleans" are actually used? The reason I ask—if 99.9999% of people say true/false, 99.9999% of people are wasting time looking through all the other options in the "false" case—and we check truthiness or falsiness very often in MSBuild. The 0.0001% of people would have to change their code, so it's clearly a breaking change, but it would be a pretty straightforward change. Just curious if we've ever done a cost-benefit analysis.

--Troublemaker

File: ref/Microsoft.Build/net/Microsoft.Build.cs (hunk 10 above)
Comment (BenVillalobos) [reply ref/Microsoft.Build/net/Microsoft.Build.cs:416]: Rebased!

File: src/Build/Instance/TaskRegistry.cs (hunk 17 above)
Comment (BenVillalobos) [reply src/Build/Instance/TaskRegistry.cs:691]: https://github.com/dotnet/msbuild/pull/6783#issuecomment-973283563

The conversation has mostly been about whether or not to log a warning, but could also be applied for logging an error. 

IMO it's not worth failing the whole build over. My thinking is the resulting build could still be valid if the user sees this warning. 

> **could** still be valid

Or is that precisely why it's worth failing the whole build over?

Comment (rainersigwald) [reply src/Build/Instance/TaskRegistry.cs:691]: My thinking is: we can always back it down to a warning if we discover a good reason to. But if we start as a warning we can't go to an error if we discover a good reason that way. So I prefer errors for new scenarios.

File: src/Build/Instance/TaskRegistry.cs (hunk 12 above)
Comment (rainersigwald) [reply src/Build/Instance/TaskRegistry.cs]: I'd like to see a profile showing that it's a significant cost before considering this further.

Comment (Forgind) [reply src/Build/Instance/TaskRegistry.cs]: I tested this, and it didn't seem nearly as impactful as I'd hoped. There may be an argument for reordering them, but I would count it very low priority.

File: src/Build/Instance/TaskRegistry.cs (hunk 16 above)
Comment (BenVillalobos) [reply src/Build/Instance/TaskRegistry.cs:647]: That's a wrinkle in this solution, at the moment it requires an exact match at the calling site. Another aspect to consider, how should this play with the overridetasks file?

File: ref/Microsoft.Build/net/Microsoft.Build.cs [hunk 18]
```diff
@@ -0,0 +1,2165 @@
+// Copyright (c) Microsoft. All rights reserved.
```
Comment (rainersigwald) [inline ref/Microsoft.Build/net/Microsoft.Build.cs:1]: Looks like this file got brought back in the merge. Delete again please?

File: src/Build/Instance/TaskRegistry.cs (hunk 16 above)
Comment (rainersigwald) [reply src/Build/Instance/TaskRegistry.cs:647]: How hard would it be to do "if the task comes from .overridetasks _and_ there's something elsewhere with an override, error"?

File: src/Build/Resources/Strings.resx (hunk 14 above)
Comment (rainersigwald) [reply src/Build/Resources/Strings.resx]: Ping on this question.

File: src/Build/Instance/TaskRegistry.cs (hunk 16 above)
Comment (BenVillalobos) [reply src/Build/Instance/TaskRegistry.cs:647]: Found an area that documentation will need to be updated (if we change it): https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-tasks?view=vs-2022#overridden-tasks

>Tasks in these files override any other tasks with the same names, including tasks in the project file.

Looks like precedence will be:
1. Project UsingTask with Override attribute
2. Anything in .overridetasks (or .tasks?)
3. Project files

> How hard would it be to do "if the task comes from .overridetasks and there's something elsewhere with an override, error"?

If we tried to go that route, one issue would be the lack of state in this class. When parsing a .tasks or .overridetasks file (see `RegisterOverrideTasks`), it calls `LoadAndRegisterFromTasksFile` which calls the static `RegisterTasksFromUsingTaskElement`. I guess we could try and determine whether we're in a .tasks or .overridetasks based on the `IFileSystem` that gets passed? Doesn't look like we can get the current file through it though.

File: src/Build/Resources/Strings.resx (hunk 14 above)
Comment (BenVillalobos) [reply src/Build/Resources/Strings.resx]: At first I figured why not have both, but notifying the build every time we picked a specific task doesn't provide much value. Especially when we can see the binary the task came from in the logs.

I'll remove the logic/resource for logging a message when an overridden task is chosen.

File: src/Build/Resources/Strings.resx (hunk 13 above)
Comment (BenVillalobos) [reply src/Build/Resources/Strings.resx]: Now that this message is gone, do you think it'd be useful to log the location when the overridetask is registered? It currently only logs the task name.

File: src/Build/Instance/TaskRegistry.cs [hunk 19]
```diff
@@ -660,7 +693,45 @@ private void RegisterTask(string taskName, AssemblyLoadInfo assemblyLoadInfo, st
...
+            if (overrideTask)
+            {
+                // Key the dictionary based on UNqualified task names
+                // This is to support partial matches on tasks like Foo.Bar and Baz.Bar
+                string unqualifiedTaskName = taskName;
+
+                if (unqualifiedTaskName.Contains('.'))
+                {
+                    unqualifiedTaskName = taskName.Split('.').Last();
+                }
```
Comment (Forgind) [inline src/Build/Instance/TaskRegistry.cs:707]:
```suggestion
                string unqualifiedTaskName = taskName.Split('.').Last();
```

File: src/Build/Instance/TaskRegistry.cs [hunk 20]
```diff
@@ -660,7 +693,45 @@ private void RegisterTask(string taskName, AssemblyLoadInfo assemblyLoadInfo, st
...
+
+                // Is the task already registered?
+                if (overriddenTasks.TryGetValue(unqualifiedTaskName, out List<RegisteredTaskRecord> recs))
+                {
+                    // check every registration that exists in the list.
+                    foreach (RegisteredTaskRecord rec in recs)
+                    {
+                        // Does the same registration already exist? (same exact name)
```
Comment (Forgind) [inline src/Build/Instance/TaskRegistry.cs:715]: Please only include useful comments. This and the past two comments added nothing to the line of code immediately below them. Comments are easily forgotten when changing code.

File: src/Build/Instance/TaskRegistry.cs [hunk 21]
```diff
@@ -660,7 +693,45 @@ private void RegisterTask(string taskName, AssemblyLoadInfo assemblyLoadInfo, st
...
+                        }
+                    }
+                    recs.Add(newRecord);
+                }
+                else
+                {
+                    // Create a dictionary containing the unqualified name (for quick lookups when the task is called).
+                    // Place the new record using the potentially fully qualified name to account for partial matches.
+                    overriddenTasks.Add(unqualifiedTaskName, new List<RegisteredTaskRecord>());
+                    overriddenTasks[unqualifiedTaskName].Add(newRecord);
```
Comment (Forgind) [inline src/Build/Instance/TaskRegistry.cs:729]:
```suggestion
                    // New record's name may be fully qualified. Use it anyway to account for partial matches.
                    List<RegisteredTaskRecord> unqualifiedTaskNameMatches = new();
                    unqualifiedTaskNameMatches.Add(newRecord);
                    overriddenTasks.Add(unqualifiedTaskName, unqualifiedTaskNameMatches);
```
?

Saves a dictionary access, too.

**Comment by rainersigwald** [comment]
> ### Prioritization order of usingtasks:
> 1. Project-level usingtask override
> 2. _.tasks/_.overridetasks file
> 3. UsingTask defined elsewhere

Is this right? Shouldn't it be

1. Project-level UsingTask override
2. `.overridetasks` file
3. Project-level UsingTask
4. `.tasks` file

?

File: src/Build.UnitTests/BackEnd/TaskRegistry_Tests.cs [hunk 22]
```diff
@@ -624,6 +628,94 @@ public void RetrieveFromCacheMatchingExactParameters_AdditionalParameters()
...
+                    };
+
+                    var request = new BuildRequestData(
+                        testProject.ProjectFile,
+                        new Dictionary<string, string>(),
+                        MSBuildConstants.CurrentToolsVersion,
+                        Array.Empty<string>(),
+                        null);
+
+                    var result = buildManager.Build(
```
Comment (rainersigwald) [inline src/Build.UnitTests/BackEnd/TaskRegistry_Tests.cs:709]: assert result true/false?

File: src/Build/Instance/TaskRegistry.cs [hunk 23]
```diff
@@ -662,7 +695,38 @@ private void RegisterTask(string taskName, AssemblyLoadInfo assemblyLoadInfo, st
...
             }
 
-            registeredTaskEntries.Add(new RegisteredTaskRecord(taskName, assemblyLoadInfo, taskFactory, taskFactoryParameters, inlineTaskRecord));
+            RegisteredTaskRecord newRecord = new RegisteredTaskRecord(taskName, assemblyLoadInfo, taskFactory, taskFactoryParameters, inlineTaskRecord);
+
+            if (overrideTask)
+            {
+                // Key the dictionary based on Unqualified task names
+                // This is to support partial matches on tasks like Foo.Bar and Baz.Bar
+                string unqualifiedTaskName = taskName.Split('.').Last();
```
Comment (rainersigwald) [inline src/Build/Instance/TaskRegistry.cs:704]: I'd say it's worth avoiding Linq for this.
```suggestion
                string[] nameComponents = taskName.Split('.');
                string unqualifiedTaskName = nameComponents[nameComponents.Length];
```

(sadly we can't use `[^1]` since we don't have `System.Index` in our net472 target)

File: src/Build/Instance/TaskRegistry.cs [hunk 24]
```diff
@@ -22,6 +22,7 @@
 using Microsoft.Build.Shared.FileSystem;
 
 using Microsoft.NET.StringTools;
+using System.Linq;
```
Comment (rainersigwald) [inline src/Build/Instance/TaskRegistry.cs:25]: Removable (see below)

File: src/Build/Resources/Strings.resx [hunk 25]
```diff
@@ -332,6 +332,13 @@
   <data name="TryingExtensionsPath" xml:space="preserve">
     <value>Trying to import {0} using extensions path {1}</value>
   </data>
+  <data name="DuplicateOverrideUsingTaskElement" xml:space="preserve">
+    <value>MSB4275: Attempted to create multiple overrides of the same task: {0}</value>
```
Comment (rainersigwald) [inline src/Build/Resources/Strings.resx:336]: Can you show an example of this message? Is it possible to point to both locations?

File: src/Build/Resources/Strings.resx [hunk 26]
```diff
@@ -332,6 +332,13 @@
...
+    <comment>{StrBegin="MSB4275: "}</comment>
+  </data>
+  <data name="OverrideUsingTaskElementCreated" xml:space="preserve">
+    <value>Created an override using task: {0}</value>
```
Comment (rainersigwald) [inline src/Build/Resources/Strings.resx:340]: Should this have the location?

**Review by rainersigwald — COMMENTED** [review]
Looking great, just a couple of things to look at!

File: src/Build/Resources/Strings.resx (hunk 25 above)
Comment (BenVillalobos) [reply src/Build/Resources/Strings.resx:336]: There doesn't seem to be a convenient way to get location from a `RegisteredTaskRecord`, which is the only reference we have to the other usingtask at the point that this is logged.

Comment (BenVillalobos) [reply src/Build/Resources/Strings.resx:336]: Here's the log after a unit test:

```
Build started.
Evaluation started ("C:\Users\bevillal\AppData\Local\Temp\pm3tsjgb.5xz\Temporary458011a3c99d422e8e03cf88a5710645\build.proj")
Created an override using task: Foo at C:\Users\bevillal\AppData\Local\Temp\pm3tsjgb.5xz\Temporary458011a3c99d422e8e03cf88a5710645\build.proj (1,89)
C:\Users\bevillal\AppData\Local\Temp\pm3tsjgb.5xz\Temporary458011a3c99d422e8e03cf88a5710645\build.proj(1,186):  error MSB4275: Attempted to create multiple overrides of the same task: Foo
Evaluation finished ("C:\Users\bevillal\AppData\Local\Temp\pm3tsjgb.5xz\Temporary458011a3c99d422e8e03cf88a5710645\build.proj")
Project "build.proj" (default targets):
Target "Bar" in project "C:\Users\bevillal\AppData\Local\Temp\pm3tsjgb.5xz\Temporary458011a3c99d422e8e03cf88a5710645\build.proj" (entry point):
Done building target "Bar" in project "build.proj".
Done building project "build.proj".
Build succeeded.
```

Comment (rainersigwald) [reply src/Build/Resources/Strings.resx:336]: This is confusing since they're on the same line but I think that's just an artifact of the test construction?

In any case there are messages for each that have a location so I'm pretty good with that.

**Review by rainersigwald — APPROVED** [review]

**Comment by Forgind** [comment]
@BenVillalobos, resolved a merge conflict in case you want to look at it, but it was pretty straightforward.
